File: dotnet/MarsRoverKata/MarsRover.cs
using System;
using System.Collections.Generic;

namespace MarsRoverKata
{
    public class MarsRover
    {
        public static class Movement
        {
            public const char Forward = 'f';
            public const char Backward = 'b';
        }
        
        private Position _position;
        private Direction _direction;

        public Position Position => _position;

        public Direction Direction => _direction;

        public MarsRover(Position position, Direction direction)
        {
            _position = position;
            _direction = direction;
        }

        public MarsRover Execute(IEnumerable<char> commands)
        {
            foreach (var command in commands)
            {
                switch (command)
                {
                    case 'r':
                        _direction = _direction.Next();
                        break;
                    case 'l':
                        _direction = _direction.Previous();
                        break;
                    case Movement.Forward:
                    case Movement.Backward:
                        Move(command);
                        break;
                    default:
                        throw new ArgumentException($"invalid command {command} provided");
                }
            }

            return this;
        }

        private void Move(char movement)
        {
            var step = movement == Movement.Forward ? 1 : -1;

            _position = _direction switch
            {
                Direction.North => new Position(_position.X, _position.Y - step),
                Direction.East => new Position(_position.X + step, _position.Y),
                Direction.South => new Position(_position.X, _position.Y + step),
                Direction.West => new Position(_position.X - step, _position.Y),
                _ => throw new InvalidOperationException($"unexpected direction value {_direction} encountered")
            };
        }
    }

    public static class DirectionExtensions
    {
        public static Direction Next(this Direction direction)
        {
            return direction switch
            {
                Direction.North => Direction.East,
                Direction.East => Direction.South,
                Direction.South => Direction.West,
                Direction.West => Direction.North,
                _ => throw new ArgumentException($"unexpected direction value {direction} provided")
            };
        }
        
        public static Direction Previous(this Direction direction)
        {
            return direction switch
            {
                Direction.North => Direction.West,
                Direction.West => Direction.South,
                Direction.South => Direction.East,
                Direction.East => Direction.North,
                _ => throw new ArgumentException($"unexpected direction value {direction} provided")
            };
        }
    }

    public enum Direction
    {
        North,
        East,
        South,
        West
    }

    public readonly struct Position
    {
        public int X { get; }

        public int Y { get; }

        public Position(int x, int y)
        {
            X = x;
            Y = y;
        }

        private bool Equals(Position other)
        {
            return X == other.X && Y == other.Y;
        }

        public override string ToString()
        {
            return $"{{ x = {X}, x = {Y} }}";
        }

        public override bool Equals(object obj)
        {
            return obj is Position other && Equals(other);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(X, Y);
        }
    }
}

File: cpp/mars_rover.cpp
#include <stdexcept>
#include "mars_rover.h"


Point::Point(int x, int y) : _x(x), _y(y) {
}


int Point::getX() const {
    return _x;
}


int Point::getY() const {
    return _y;
}


bool Point::operator==(const Point &rhs) const {
    return _x == rhs._x &&
           _y == rhs._y;
}


bool Point::operator!=(const Point &rhs) const {
    return !(rhs == *this);
}


Direction getNextDirection(Direction direction) {
    switch (direction) {
    case Direction::East:
        return Direction::North;
    case Direction::North:
        return Direction::West;
    case Direction::West:
        return Direction::South;
    case Direction::South:
        return Direction::East;
    default:
        throw std::logic_error(
                std::string("unexpected direction value ") + std::to_string(static_cast<int>(direction)));
    }
}


Direction getPreviousDirection(Direction direction) {
    switch (direction) {
    case Direction::East:
        return Direction::South;
    case Direction::North:
        return Direction::East;
    case Direction::West:
        return Direction::North;
    case Direction::South:
        return Direction::West;
    default:
        throw std::logic_error(
                std::string("unexpected direction value ") + std::to_string(static_cast<int>(direction)));
    }
}


MarsRover::MarsRover(const Point &position, Direction direction) : _position(position), _direction(direction) {}


const Point &MarsRover::getPosition() const {
    return _position;
}


Direction MarsRover::getDirection() const {
    return _direction;
}


MarsRover &MarsRover::execute(const std::string &commands) {
    for (auto c : commands) {
        switch (c) {
        case static_cast<decltype(c)>(Rotation::Left):
            rotateLeft();
            break;
        case static_cast<decltype(c)>(Rotation::Right):
            rotateRight();
            break;
        case static_cast<decltype(c)>(Movement::Forward):
        case static_cast<decltype(c)>(Movement::Backward):
            move(static_cast<Movement>(c));
            break;
        default:
            throw std::logic_error(
                    std::string("unknown command '") + c + "'");
        }
    }
    return *this;
}


void MarsRover::move(Movement movement) {

    auto step = movement == Movement::Forward ? 1 : -1;

    switch (_direction) {
    case Direction::East:
        _position = Point(_position.getX() + step, _position.getY());
        break;
    case Direction::North:
        _position = Point(_position.getX(), _position.getY() - step);
        break;
    case Direction::West:
        _position = Point(_position.getX() - step, _position.getY());
        break;
    case Direction::South:
        _position = Point(_position.getX(), _position.getY() + step);
        break;
    default:
        throw std::logic_error(
                std::string("unexpected direction value ") + std::to_string(static_cast<int>(_direction)));
    }
}


void MarsRover::rotateLeft() {
    _direction = getNextDirection(_direction);
}


void MarsRover::rotateRight() {
    _direction = getPreviousDirection(_direction);
}

